File: 75.颜色分类.py
#
# @lc app=leetcode.cn id=75 lang=python3
#
# [75] 颜色分类
#

# @lc code=start
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # nums.sort(key = lambda e: e)
        def swap2Position(nums, target, index):
            for i in range(index, len(nums)):
                if nums[i] == target:
                    nums[index], nums[i] = nums[i], nums[index]
                    index += 1
            return index
        index = swap2Position(nums, 0, 0)  # 先移动 0
        swap2Position(nums, 1, index)  # 再移动 1， 2 自然就在最后连
        # 不能swap2Position(nums[index:], 1) 因为传递过去的会是一个拷贝
# @lc code=end

File: 75.颜色分类.go
/*
 * @lc app=leetcode.cn id=75 lang=golang
 *
 * [75] 颜色分类
 */

// @lc code=start
func sortColors(nums []int) {
	// sort.Slice(nums, func(i, j int) bool {
	//     return nums[i] < nums[j]
	// })
	index := swap2Position(nums, 0)
	swap2Position(nums[index:], 1)
}

func swap2Position(nums []int, target int) (count int) {
	for idx, num := range nums {
		if num == target {
			nums[count], nums[idx] = nums[idx], nums[count]
			count++
		}
	}
	return
}

// @lc code=end

